# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/2d/DrawTargetCairo.cpp
# Commit: d8b66c3db775
# Full Hash: d8b66c3db77594a0e43de8f93fb925765767031c
# Author: Steven Michaud <smichaud@pobox.com>
# Date: 2022-02-04 03:46:27
# Description:
#   Bug 1719215 - Constrain clipping in CreateSubImageForData(). r=lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D137777
# ==============================================================================

diff -r 12f866a5b267 -r d8b66c3db775 gfx/2d/DrawTargetCairo.cpp
--- a/gfx/2d/DrawTargetCairo.cpp	Thu Feb 03 18:53:13 2022 +0000
+++ b/gfx/2d/DrawTargetCairo.cpp	Thu Feb 03 18:54:01 2022 +0000
@@ -254,6 +254,15 @@
   return nullptr;
 }
 
+// We're creating a subimage from the parent image's data (in aData) without
+// altering that data or its stride. This constrains the values in aRect, and
+// how they're used. Callers must see to it that the parent fully contains the
+// subimage. Here we ensure that no clipping is done in the X dimension at the
+// beginning of any line. (To do otherwise would require creating a copy of
+// aData from parts of every line in aData (from aRect.Y() to aRect.Height()),
+// and setting the copy to a different stride.) A non-zero aRect.X() is used
+// only to specify the subimage's location in its parent (via
+// cairo_surface_set_device_offset()). This change resolves bug 1719215.
 static cairo_surface_t* CreateSubImageForData(unsigned char* aData,
                                               const IntRect& aRect, int aStride,
                                               SurfaceFormat aFormat) {
@@ -261,12 +270,12 @@
     gfxWarning() << "DrawTargetCairo.CreateSubImageForData null aData";
     return nullptr;
   }
-  unsigned char* data =
-      aData + aRect.Y() * aStride + aRect.X() * BytesPerPixel(aFormat);
+  unsigned char* data = aData + aRect.Y() * aStride;
 
   cairo_surface_t* image = cairo_image_surface_create_for_data(
       data, GfxFormatToCairoFormat(aFormat), aRect.Width(), aRect.Height(),
       aStride);
+  // Set the subimage's location in its parent
   cairo_surface_set_device_offset(image, -aRect.X(), -aRect.Y());
   return image;
 }
